Motech openmrs concepts

dannyroberts opened the pull request:
<img width="1436" alt="screen shot 2017-04-17 at 3 04 48 pm" src="https://cloud.githubusercontent.com/assets/137212/25090584/5b3e5c7e-2384-11e7-959e-ffbf1981a607.png">

I'd review commit by commit. Note that one of the early commits is moving code over from the other repo.

dannyroberts committed:
add placeholder motech OpenMRS concepts page

dannyroberts committed:
copy some openmrs starter code over from motechhub repo

dannyroberts committed:
make migration for openmrs

dannyroberts committed:
tweak openmrs concept views

dannyroberts committed:
connceted_accounts => connected_accounts

dannyroberts committed:
add openmrs concepts page

and hidden url to trigger sync of concepts from openmrs

dannyroberts committed:
change primary key on OpenmrsConcept

dannyroberts committed:
show all concepts that are not used as answers

instead of just concepts that have answers

dannyroberts committed:
add button to sync openmrs concepts

dimagimon commented on corehq/apps/motech/static/motech/js/motech_openmrs_concepts.js:
```diff
@@ -0,0 +1,33 @@
+                self.concepts([]);
+                self.syncButton.inProgress(true);
+                self.syncButton.hasError(false);
+                $.get(urllib.reverse('openmrs_sync_concepts')).done(function (data) {
```
'data' is defined but never used (no-unused-vars)

dimagimon commented on corehq/apps/motech/static/motech/js/motech_openmrs_concepts.js:
```diff
@@ -0,0 +1,33 @@
+                    self.syncButton.inProgress(false);
+                    self.syncButton.hasError(true);
+                });
+            }
```
Missing trailing comma. (comma-dangle)

dimagimon commented on corehq/apps/motech/openmrs/concepts/sync.py:
```diff
@@ -0,0 +1,113 @@
+import json
+from django.db import IntegrityError
```
F401 'IntegrityError' imported but unused

dimagimon commented on corehq/apps/motech/openmrs/tests.py:
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
F401 'TestCase' imported but unused

dannyroberts committed:
[ci skip] fix linter errors

dannyroberts committed:
remove unused view

benrudolph commented on corehq/apps/motech/openmrs/concepts/views.py:
```diff
@@ -0,0 +1,50 @@
+        all_openmrs_concepts = OpenmrsConcept.objects.filter(Q(account=account) & ~Q(answers=None))
+        openmrs_concepts = all_openmrs_concepts.filter(names__icontains=search)
+        first_50 = openmrs_concepts[:50]
+        return JsonResponse({'concepts': [
```
does jsonresponse not already automatically call `to_json` on all objects?

dannyroberts replied:
It doesn't; `JsonResponse` is a Django built-in that takes in a `dict` and returns a Django `HttpResponse` with the desired status code, a JSON body based on your dictionary, and content type set to the correct JSON mimetype (`application/json`). `.to_json` is a `JsonObject` method that converts from an instance of a `JsonObject` (subclass) to a JSON-serializable `dict` (not actually to JSON). I think it's a little confusing because at some point, possibly before Django's `JsonResponse` existed, we had defined a `json_response` helper that used our custom JSON handler that actually _does_ call the `.to_json()` method on the object-to-be-serialized, if it exists. This method is still used in some places, but I now prefer using the built-in `JsonResponse`. (Sorry for the characteristic over-explanation here!)

benrudolph commented on corehq/apps/motech/templates/motech/motech_openmrs_concepts.html:
```diff
@@ -0,0 +1,47 @@
+            disable: syncButton.inProgress,
+            click: syncButton.sync
+        ">
+            <!--ko ifnot: syncButton.inProgress-->Sync from OpenMRS<!--/ko-->
```
not sure if these need to be translate, but flaggin in case

dannyroberts replied:
That's fair. I was in kind of "this isn't really going to be live yet so it doesn't matter" mode, but you're right—it's way easier to do this kind of thing as you go along than it is to do them in bulk after the fact.

- [ ] Translate all easily-translatable strings in this PR's diff
- [ ] Add a todo comment next to any hard-to-translate string that should eventually be translated (if any).

proteusvacuum commented on corehq/apps/motech/openmrs/concepts/sync.py:
```diff
@@ -0,0 +1,112 @@
+
+        if len(answers_relationships) == len(delayed):
+            # this is going to be an infinite loop
+            raise Exception(delayed)
```
nit: do you want to have a better named exception?

dannyroberts replied:
OK fine :)

- [ ] Raise a better exception than `Exception`

dannyroberts replied:
Actually I think this approach isn't necessary at all—at one point I think I was trying to get each concept to save only once, so I basically had to feel out the dependency graph using that method. Now that they're all saved up front and then they're re-saved after adding answers (which may only actually update the concept-to-answer table) I think it's no longer necessary.

proteusvacuum commented on corehq/apps/motech/openmrs/concepts/sync.py:
```diff
@@ -0,0 +1,112 @@
+            answers_relationships.append((concept, answers))
+        concept.save()
+
+    while answers_relationships:
```
I don't know what OpenMRS concepts are. Can you explain what this loop is doing? it looks to me as though it just looks in the database to see if all the `answers` are the same as the ones you received from the API. Is that correct? I might just be a bit slow this afternoon.

dannyroberts replied:
Concepts correspond to a few different things as we think of them in CommCare, including
- questions
- answers
- case properties (e.g. patient properties)
- case property values (e.g. patient property values)

A concept that is... _conceptually_... a "question", will be defined with "answers" which are other concepts.

The key lines in here is

```python
concept.answers = answer_concepts
concept.save()
```

Conceptually this is really just part of the the previous loop—for every concept that comes in you want to save the corresponding answers in the question & answer table, which is a many-to-many table from concepts (questions) to concepts (answers). But you can't save to that table until both the question and answer involved in a row has been saved. Hence, I defer populating that table until each concept has been saved.

proteusvacuum commented on corehq/apps/motech/openmrs/patients/views.py:
```diff
@@ -0,0 +1,43 @@
+        json.dumps([{
+            'uuid': p['uuid'],
+            'display': p['display'],
+        } for p in patients]), content_type='text/json')
```
I think the content type for json is `application/json`.

dannyroberts replied:
True

proteusvacuum commented on corehq/apps/motech/openmrs/patients/views.py:
```diff
@@ -0,0 +1,43 @@
+        openmrs_patient_identifier_type_json_from_api_json(patient_identifier_type).to_json()
+        for patient_identifier_type in api.get_all()
+    ]
+    return HttpResponse(json.dumps(patient_identifier_types), content_type='text/json')
```
`application/json`

proteusvacuum commented on corehq/apps/motech/openmrs/patients/views.py:
```diff
@@ -0,0 +1,43 @@
+            'badrequest': 'Must include params idtype and id'
+        }), status=400)
+    requests = get_openmrs_requests_object(domain)
+    patients = requests.get('/openmrs/ws/rest/v1/patient', params={'q': patient_id, 'v': 'full'}).json()['results']
```
what if this fails?

proteusvacuum commented on corehq/apps/motech/openmrs/urls.py:
```diff
@@ -0,0 +1,23 @@
+        views.sync_concepts,
+        name='openmrs_sync_concepts'),
+
+    # patients
```
are these ever called?

dannyroberts replied:
I think you're right and I can just get rid of the whole `patients` portion. It's copied from an earlier prototype that did use it, but I can just add it back once it becomes a thing.

proteusvacuum commented on corehq/apps/motech/openmrs/urls.py:
```diff
@@ -0,0 +1,23 @@
+        name='all_openmrs_concepts'),
+    url(r'^rest/concept/search/$',
+        views.concept_search,
+        name='openmrs_concept_search'),
```
where is this called from?

dannyroberts replied:
It's not actually called here yet, but it's setting up an endpoint that pages can use in an async call. Tempted to just leave it here, since I'm planning to use it soon.

dannyroberts committed:
Merge branch 'master' into motech-openmrs-concepts

dannyroberts committed:
add i18n to motech_openmrs_concepts.html

dannyroberts committed:
clean up sync_concepts_from_openmrs

can use a simpler approach now that concepts are saved up front
rather than only after adding answers

dannyroberts committed:
remove openmrs patient-related views (unused)

dannyroberts commented:
@proteusvacuum updated

proteusvacuum commented:
Motech! Gotech!
